okio/src/jvmMain/kotlin/okio/internal/ResourceFileSystem.kt

```diff
@@ -136,4 +112,56 @@ internal class ResourceFileSystem internal constructor(
     throw IOException("$this is read-only")
 
   override fun delete(path: Path): Unit = throw IOException("$this is read-only")
+
+  private fun Path.toRelativePath(): String = canonicalize(this).toString().substring(1)
+
+  private companion object {
+    val ROOT = "/".toPath()
+
+    fun Path.removeBase(base: Path): Path {
+      val prefix = base.toString()
+      return ROOT / (toString().removePrefix(prefix).replace('\\', '/'))
+    }
+
+    /**
+     * Returns a search path of classpath roots. Each element contains a file system to use, and
+     * the base directory of that file system to search from.
+     */
+    fun ClassLoader.toClasspathRoots(): List<Pair<FileSystem, Path>> {
+      // We'd like to build this upon an API like ClassLoader.getURLs() but unfortunately that
+      // API exists only on URLClassLoader (and that isn't the default class loader implementation).
+      //
+      // The closest we have is `ClassLoader.getResources("")`. It returns all classpath roots that
+      // are directories but none that are .jar files. To mitigate that we also search for all
+      // `META-INF/MANIFEST.MF` files, hastily assuming that every .jar file will have such an
+      // entry.
+      //
+      // Classpath entries that aren't directories and don't have a META-INF/MANIFEST.MF file will
+      // not be visible in this file system.
+      return getResources("").toList().mapNotNull { it.toFileRoot() } +
+        getResources("META-INF/MANIFEST.MF").toList().mapNotNull { it.toJarRoot() }
+    }
+
+    fun URL.toFileRoot(): Pair<FileSystem, Path>? {
+      if (protocol != "file") return null // Ignore unexpected URLs.
```

Comment:
I wonder what this does in Graal native images...?

Reply:
If it doesn't behave we should probably add specific support via nativeImage build time generation of file lists.  So not a blocker here.  That would be a fun feature to add, and potentially fix generally or with some specific API for that world.

Reply:
Yep, great idea. It'll likely be different from this! Gotta look at how the public suffix list works in OkHttp?!

Reply:
We whitelist this https://github.com/square/okhttp/blob/master/okhttp/src/main/resources/META-INF/native-image/okhttp/okhttp/resource-config.json